README.md
# Readme
![repo-size](https://img.shields.io/github/repo-size/Py-KMS-Organization/py-kms)
![open-issues](https://img.shields.io/github/issues/Py-KMS-Organization/py-kms)
![last-commit](https://img.shields.io/github/last-commit/Py-KMS-Organization/py-kms/master)
![docker-pulls](https://img.shields.io/docker/pulls/pykmsorg/py-kms)
![read-the-docs](https://img.shields.io/readthedocs/py-kms)
***

_Keep in mind that this project is not intended for production use. Feel free to use it to test your own systems or maybe even learn something from the protocol structure. :)_

## History
_py-kms_ is a port of node-kms created by [cyrozap](http://forums.mydigitallife.info/members/183074-markedsword), which is a port of either the C#, C++, or .NET implementations of KMS Emulator. The original version was written by [CODYQX4](http://forums.mydigitallife.info/members/89933-CODYQX4) and is derived from the reverse-engineered code of Microsoft's official KMS.
This version of _py-kms_ is for itself a fork of the original implementation by [SystemRage](https://github.com/SystemRage/py-kms), which was abandoned early 2021.

## Features
- Responds to `v4`, `v5`, and `v6` KMS requests.
- Supports activating:
	- Windows Vista 
	- Windows 7 
	- Windows 8
	- Windows 8.1
	- Windows 10 ( 1511 / 1607 / 1703 / 1709 / 1803 / 1809 )
    - Windows 10 ( 1903 / 1909 / 20H1, 20H2, 21H1, 21H2 )
    - Windows 11 ( 21H2 )
	- Windows Server 2008
	- Windows Server 2008 R2
	- Windows Server 2012
	- Windows Server 2012 R2
	- Windows Server 2016
	- Windows Server 2019
	- Windows Server 2022
	- Microsoft Office 2010 ( Volume License )
	- Microsoft Office 2013 ( Volume License )
	- Microsoft Office 2016 ( Volume License )
	- Microsoft Office 2019 ( Volume License )
	- Microsoft Office 2021 ( Volume License )
  - It's written in Python (tested with Python 3.10.1).
  - Supports execution by `Docker`, `systemd` and many more...
  - Uses `sqlite` for persistent data storage (with a simple web-based explorer).

## Documentation
The wiki has been completly reworked and is now available on [readthedocs.com](https://py-kms.readthedocs.io/en/latest/). It should you provide all necessary information how to setup and to use _py-kms_ , all without clumping this readme. The documentation also houses more details about activation with _py-kms_ and how to get GVLK keys.
       
## Quick start
- To start the server, execute `python3 pykms_Server.py [IPADDRESS] [PORT]`, the default _IPADDRESS_ is `0.0.0.0` ( all interfaces ) and the default _PORT_ is `1688`. Note that both the address and port are optional. It's allowed to use IPv4 and IPv6 addresses. If you have a IPv6-capable dual-stack OS, a dual-stack socket is created when using a IPv6 address.
- To start the server automatically using Docker, execute `docker run -d --name py-kms --restart always -p 1688:1688 ghcr.io/py-kms-organization/py-kms`.
- To show the help pages type: `python3 pykms_Server.py -h` and `python3 pykms_Client.py -h`.

## License
   - _py-kms_ is [![Unlicense](https://img.shields.io/badge/license-unlicense-lightgray.svg)](https://github.com/SystemRage/py-kms/blob/master/LICENSE)
   - _py-kms GUI_ is [![MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://github.com/SystemRage/py-kms/blob/master/LICENSE.gui.md) © Matteo ℱan


CHANGELOG.md
# Changelog

### py-kms_2022-12-16
- Added support for new web-gui into Docker
- Implemented whole-new web-based GUI with Flask
- Removed old GUI (Etrigan) from code and resources
- Removed sqliteweb
- Removed Etrigan (GUI)

### py-kms_2022-12-07
- Added warning about Etrigan (GUI) being deprecated
- More docs (do not run on same machine as client)
- Added Docker support for multiple listen IPs
- Added graceful Docker shutdowns

### py-kms_2021-12-23
- More Windows 10/11 keys
- Fixed some deprecation warnings
- Fixed SO_REUSEPORT platform checks
- Fixed loglevel "MININFO" with Docker
- Added Docker healthcheck
- Added UID/GID change support for Docker
- Dependabot alerts

### py-kms_2021-10-22
- Integrated Office 2021 GLVK keys & database
- Docker entrypoint fixes
- Updated docs to include SQLite stuff
- Fix for undefined timezones
- Removed LOGFILE extension checks
- Added support for Windows 11

### py-kms_2021-10-07
- Helm charts for Kubernetes deployment
- Windows 2022 updates
- Faster Github Action builds

### py-kms_2021-11-12
- Addded GHCR support
- Docs table reformatted
- Updated GUI
- Windows Sandbox fix
- Added contribution guidelines
- Docker multiarch
- Reshot screenshots in docs

### py-kms_2020-10-01
- Sql database path customizable.
- Sql database file keeps different AppId.
- Support for multi-address connection.
- Added timeout send / receive.

### py-kms_2020-07-01
- py-kms Gui: now matches all cli options, added modes onlyserver / onlyclient,
  added some animations.
- Added suboptions FILEOFF and STDOUTOFF of -F.
- Created option for asynchronous messages.
- Cleaned options parsing process.

### py-kms_2020-02-02
- Optimized pretty-print messages process.
- Added -F FILESTDOUT option.
- Added deamonization options (via [Etrigan](https://github.com/SystemRage/Etrigan) project).
- py-kms GUI resurrected (and improved).
- Cleaned, cleaned, cleaned.

### py-kms_2019-05-15
- Merging for Python2 / Python3 compatibility all-in-one.
- Added new options: 
    - timeout, [logsize](https://github.com/SystemRage/py-kms/pull/21).
    - more control on logging and info visualization (custom loglevel and stdout logfile redirection) to match [this](https://github.com/SystemRage/py-kms/issues/22) request.
- Setup for multithreading support.
- Graphical improvements:
    - added a (_"really silly"_) tkinter GUI as an alternative to command line.
- [Dockerized](https://github.com/SystemRage/py-kms/pull/20) with sqlite-web. 
- Fixed activation threshold.
- Renamed files, cosmetics and many other little big adjustments.

### py-kms_2018-11-15
 - Implemented some good modifications inspired by [this](https://github.com/ThunderEX/py-kms) other fork.
 	- Clean up code ( deleted no longer useful files randomHWID.py, randomEPID.py, timezones.py;
			erased useless functions and import modules )
	- Reading parameters directly from a slightly modified KmsDataBase.xml ( created with LicenseManager 5.0 by Hotbird64 HGM ) with kmsDB2Dict.py
- Added support for Windows Server 2019 and Office 2019.
- Improved random EPID generation.
- Corrected [this](https://github.com/SystemRage/py-kms/issues/8) in kmsBase.py

### py-kms_2018-03-01
 - *py-kms NOW is for Python3 too ( py3-kms ), the previous one ( written with Python2 ) is renamed py2-kms*
 - *Repaired logging messages*
 - *Added pretty processing messages*
 
### py-kms_2017-06-01
 - *Added option verbose logging in a file*
 - *Updated "kmsBase.py" with new SKUIDs*
 - *Added a brief guide "py-kms-Guide.pdf" ( replaced "client-activation.txt" )*
 - *Added a well formatted and more complete list of volume keys "py-kms-ClientKeys.pdf" ( replaced "client-keys.txt" )*

### py-kms_2016-12-30
 - *Updated kmsBase.py (Matches LicenseManager 4.6.0 by Hotbird64 HGM)*

### py-kms_2016-08-13
 - *Fixed major bug on Response function*
 - *Fixed Random PID Generator (thanks: mkuba50)*

### py-kms_2016-08-12
 - *Added missing UUID, credits: Hotbird64*
 - *Added Windows Server 2016 in random PID generator*
		
### py-kms_2016-08-11
 - *Added Windows Server 2016 UUID*
 - *Fixed GroupID and PIDRange*
 - *Added Office 2016 CountKMSID*
	
### py-kms_2015-07-29
 - *Added Windows 10 UUID*
		
### py-kms_2014-10-13 build 3:
 - *Added Client Activation Examples: "client-activation.txt"*
 - *Added Volume Keys: "client-keys.txt"*

### py-kms_2014-10-13 build 2:
 - *Added missing skuIds in file "kmsbase.py". Thanks (user_hidden)*
	
### py-kms_2014-10-13 build 1:
 - *The server now outputs the hwid in use.*
 - *The server hwid can be random by using parameter: "-w random". Example: "python server.py -w random"*
 - *Included file "randomHWID.py" to generate random hwid on demand.*
 - *Included file "randomPID.py" to generate random epid and hwid on demand.*

### py-kms_2014-03-21T232943Z:
- *The server HWID can now be specified on the command line.*
- *The client will print the HWID when using the v6 protocol.*
	
### py-kms_2014-01-03T032458Z:
 - *Made the sqlite3 module optional.*
 - *Changed the "log" flag to an "sqlite" flag and made a real log flag in preparation for when real request logging is implemented.*

### py-kms_2014-01-03T025524Z:
 - *Added RPC response decoding to the KMS client emulator.*
		
### py-kms_2013-12-30T064443Z:
 - *The v4 hash now uses the proper pre-expanded key.*

### py-kms_2013-12-28T073506Z:
 - *Modified the v4 code to use the custom aes module in order to make it more streamlined and efficient.*

### py-kms_2013-12-20T051257Z:
 - *Removed the need for the pre-computed table (tablecomplex.py) for v4 CMAC calculation, cutting the zip file size in half.*

### py-kms_2013-12-16T214638Z:
 - *Switched to getting the to-be-logged request time from the KMS server instead of the client.*

### py-kms_2013-12-16T030001Z:
 - *You can now specify the CMID and the Machine Name to use with the client emulator.*

### py-kms_2013-12-16T021215Z:

client.py
import argparse
import binascii
import datetime
import random
import socket
import string
import struct
import sys
import uuid
import filetimes, rpcBind, rpcRequest
from dcerpc import MSRPCHeader, MSRPCBindNak, MSRPCRequestHeader, MSRPCRespHeader
from kmsBase import kmsBase, UUID
from kmsRequestV4 import kmsRequestV4
from kmsRequestV5 import kmsRequestV5
from kmsRequestV6 import kmsRequestV6
from rpcBase import rpcBase
import logging
import os

config = {}

def main():
	parser = argparse.ArgumentParser()
	parser.add_argument("ip", action="store", help='The IP address or hostname of the KMS server.', type=str)
	parser.add_argument("port", nargs="?", action="store", default=1688,
                            help='The port the KMS service is listening on. The default is \"1688\".', type=int)
	parser.add_argument("-m", "--mode", dest="mode",
                            choices=["WindowsVista","Windows7","Windows8","Windows81","Windows10","Office2010","Office2013","Office2016"], default="Windows81",
                            help='Use this flag to manually specify a Microsoft product for testing the server. The default is \"Windows81\".', type=str)
	parser.add_argument("-c", "--cmid", dest="cmid", default=None,
                            help='Use this flag to manually specify a CMID to use. If no CMID is specified, a random CMID will be generated.', type=str)
	parser.add_argument("-n", "--name", dest="machineName", default=None,
                            help='Use this flag to manually specify an ASCII machineName to use. If no machineName is specified,\
a random machineName will be generated.', type=str)
	parser.add_argument("-v", "--loglevel", dest="loglevel", action="store", default="ERROR", choices=["CRITICAL", "ERROR", "WARNING", "INFO", "DEBUG"],
                            help='Use this flag to set a Loglevel. The default is \"ERROR\".', type=str)
        parser.add_argument("-f", "--logfile", dest="logfile", action="store", default=os.path.dirname(os.path.abspath( __file__ )) + "/pykms_client.log",
                            help='Use this flag to set an output Logfile. The default is \"pykms_client.log\".', type=str)
        
	config.update(vars(parser.parse_args()))

	logging.basicConfig(level=config['loglevel'], format='%(asctime)s %(levelname)-8s %(message)s',
                            datefmt='%a, %d %b %Y %H:%M:%S', filename=config['logfile'], filemode='w')

	checkConfig()
	config['call_id'] = 1
	updateConfig()
	s = socket.socket()
	logging.info("Connecting to %s on port %d..." % (config['ip'], config['port']))
	s.connect((config['ip'], config['port']))
	logging.info("Connection successful !")
	binder = rpcBind.handler(None, config)
	RPC_Bind = str(binder.generateRequest())
	logging.info("Sending RPC bind request...")
	s.send(RPC_Bind)
	try:
		bindResponse = s.recv(1024)
	except socket.error, e:
		if e[0] == 104:
			logging.error("Connection reset by peer. Exiting...")
			sys.exit()
		else:
			raise
	if bindResponse == '' or not bindResponse:
		logging.error("No data received ! Exiting...")
		sys.exit()
	packetType = MSRPCHeader(bindResponse)['type']
	if packetType == rpcBase.packetType['bindAck']:
		logging.info("RPC bind acknowledged.")
		kmsRequest = createKmsRequest()
		requester = rpcRequest.handler(kmsRequest, config)
		s.send(str(requester.generateRequest()))
		response = s.recv(1024)
		logging.debug("Response: %s" % binascii.b2a_hex(response))
		parsed = MSRPCRespHeader(response)
		kmsData = readKmsResponse(parsed['pduData'], kmsRequest, config)
		kmsResp = kmsData['response']
		
		try:
			hwid = kmsData['hwid']
		except:
			hwid = None
		logging.info("KMS Host ePID: %s" % kmsResp['kmsEpid'].decode('utf-16le').encode('utf-8'))
		if hwid is not None:
                        logging.info("KMS Host HWID: %s" % binascii.b2a_hex(hwid).upper())
                        
		logging.info("KMS Host Current Client Count: %s" % kmsResp['currentClientCount'])
		logging.info("KMS VL Activation Interval: %s" % kmsResp['vLActivationInterval'])
		logging.info("KMS VL Renewal Interval: %s" % kmsResp['vLRenewalInterval'])
		
	elif packetType == rpcBase.packetType['bindNak']:
		logging.info(MSRPCBindNak(bindResponse).dump())
		sys.exit()
	else:
		logging.critical("Something went wrong.")
		sys.exit()
		

def checkConfig():
	if config['cmid'] is not None:
		try:
			uuid.UUID(config['cmid'])
		except:
                        logging.error("Bad CMID. Exiting...")
			sys.exit()
	if config['machineName'] is not None:
		if len(config['machineName']) < 2 or len(config['machineName']) > 63:
                        logging.error("machineName must be between 2 and 63 characters in length.")
			sys.exit()

def updateConfig():
	if config['mode'] == 'WindowsVista':
		config['RequiredClientCount'] = 25
		config['KMSProtocolMajorVersion'] = 4
		config['KMSProtocolMinorVersion'] = 0
		config['KMSClientLicenseStatus'] = 2
		config['KMSClientAppID'] = "55c92734-d682-4d71-983e-d6ec3f16059f"
		config['KMSClientSkuID'] = "cfd8ff08-c0d7-452b-9f60-ef5c70c32094"
		config['KMSClientKMSCountedID'] = "212a64dc-43b1-4d3d-a30c-2fc69d2095c6"
	elif config['mode'] == 'Windows7':
		config['RequiredClientCount'] = 25
		config['KMSProtocolMajorVersion'] = 4
		config['KMSProtocolMinorVersion'] = 0
		config['KMSClientLicenseStatus'] = 2
		config['KMSClientAppID'] = "55c92734-d682-4d71-983e-d6ec3f16059f"
		config['KMSClientSkuID'] = "ae2ee509-1b34-41c0-acb7-6d4650168915"
		config['KMSClientKMSCountedID'] = "7fde5219-fbfa-484a-82c9-34d1ad53e856"
	elif config['mode'] == 'Windows8':
		config['RequiredClientCount'] = 25
		config['KMSProtocolMajorVersion'] = 5
		config['KMSProtocolMinorVersion'] = 0
		config['KMSClientLicenseStatus'] = 2
		config['KMSClientAppID'] = "55c92734-d682-4d71-983e-d6ec3f16059f"
		config['KMSClientSkuID'] = "458e1bec-837a-45f6-b9d5-925ed5d299de"
		config['KMSClientKMSCountedID'] = "3c40b358-5948-45af-923b-53d21fcc7e79"
	elif config['mode'] == 'Windows81':
		config['RequiredClientCount'] = 25
		config['KMSProtocolMajorVersion'] = 6
		config['KMSProtocolMinorVersion'] = 0
		config['KMSClientLicenseStatus'] = 2
		config['KMSClientAppID'] = "55c92734-d682-4d71-983e-d6ec3f16059f"
		config['KMSClientSkuID'] = "81671aaf-79d1-4eb1-b004-8cbbe173afea"
		config['KMSClientKMSCountedID'] = "cb8fc780-2c05-495a-9710-85afffc904d7"
	elif config['mode'] == 'Windows10':
		config['RequiredClientCount'] = 25
		config['KMSProtocolMajorVersion'] = 6
		config['KMSProtocolMinorVersion'] = 0
		config['KMSClientLicenseStatus'] = 2
		config['KMSClientAppID'] = "55c92734-d682-4d71-983e-d6ec3f16059f"
		config['KMSClientSkuID'] = "73111121-5638-40f6-bc11-f1d7b0d64300"
		config['KMSClientKMSCountedID'] = "58e2134f-8e11-4d17-9cb2-91069c151148"
	elif config['mode'] == 'Office2010':
		config['RequiredClientCount'] = 5
		config['KMSProtocolMajorVersion'] = 4
		config['KMSProtocolMinorVersion'] = 0
		config['KMSClientLicenseStatus'] = 2
		config['KMSClientAppID'] = "59a52881-a989-479d-af46-f275c6370663"
		config['KMSClientSkuID'] = "6f327760-8c5c-417c-9b61-836a98287e0c"
		config['KMSClientKMSCountedID'] = "e85af946-2e25-47b7-83e1-bebcebeac611"
	elif config['mode'] == 'Office2013':
		config['RequiredClientCount'] = 5
		config['KMSProtocolMajorVersion'] = 5
		config['KMSProtocolMinorVersion'] = 0
		config['KMSClientLicenseStatus'] = 2
		config['KMSClientAppID'] = "0ff1ce15-a989-479d-af46-f275c6370663"
		config['KMSClientSkuID'] = "b322da9c-a2e2-4058-9e4e-f59a6970bd69"
		config['KMSClientKMSCountedID'] = "e6a6f1bf-9d40-40c3-aa9f-c77ba21578c0"
	elif config['mode'] == 'Office2016':
		config['RequiredClientCount'] = 5
		config['KMSProtocolMajorVersion'] = 6
		config['KMSProtocolMinorVersion'] = 0
		config['KMSClientLicenseStatus'] = 2
		config['KMSClientAppID'] = "0ff1ce15-a989-479d-af46-f275c6370663"
		config['KMSClientSkuID'] = "d450596f-894d-49e0-966a-fd39ed4c4c64"
		config['KMSClientKMSCountedID'] = "85b5f61b-320b-4be3-814a-b76b2bfafc82"

def createKmsRequestBase():
	requestDict = kmsBase.kmsRequestStruct()
	requestDict['versionMinor'] = config['KMSProtocolMinorVersion']
	requestDict['versionMajor'] = config['KMSProtocolMajorVersion']
	requestDict['isClientVm'] = 0
	requestDict['licenseStatus'] = config['KMSClientLicenseStatus']
	requestDict['graceTime'] = 43200
	requestDict['applicationId'] = UUID(uuid.UUID(config['KMSClientAppID']).bytes_le)
	requestDict['skuId'] = UUID(uuid.UUID(config['KMSClientSkuID']).bytes_le)
	requestDict['kmsCountedId'] = UUID(uuid.UUID(config['KMSClientKMSCountedID']).bytes_le)
	requestDict['clientMachineId'] = UUID(uuid.UUID(config['cmid']).bytes_le if (config['cmid'] is not None) else uuid.uuid4().bytes_le)
	requestDict['previousClientMachineId'] = '\0' * 16 #requestDict['clientMachineId'] # I'm pretty sure this is supposed to be a null UUID.
	requestDict['requiredClientCount'] = config['RequiredClientCount']
	requestDict['requestTime'] = filetimes.dt_to_filetime(datetime.datetime.utcnow())
	requestDict['machineName'] = (config['machineName'] if (config['machineName'] is not None) else ''.join(random.choice(string.letters + string.digits) for i in range(random.randint(2,63)))).encode('utf-16le')
	requestDict['mnPad'] = '\0'.encode('utf-16le') * (63 - len(requestDict['machineName'].decode('utf-16le')))

	# Debug Stuff
	logging.debug("Request Base Dictionary: %s" % requestDict.dump())

	return requestDict

def createKmsRequest():
	# Update the call ID
	config['call_id'] += 1

	# KMS Protocol Major Version
	if config['KMSProtocolMajorVersion'] == 4:
		handler = kmsRequestV4(None, config)
	elif config['KMSProtocolMajorVersion'] == 5:
		handler = kmsRequestV5(None, config)
	elif config['KMSProtocolMajorVersion'] == 6:
		handler = kmsRequestV6(None, config)
	else:
		return None

	requestBase = createKmsRequestBase()
	return handler.generateRequest(requestBase)

def readKmsResponse(data, request, config):
	if config['KMSProtocolMajorVersion'] == 4:
		logging.info("Received V4 response")
		response = readKmsResponseV4(data, request)
	elif config['KMSProtocolMajorVersion'] == 5:
		logging.info("Received V5 response")
		response = readKmsResponseV5(data)
	elif config['KMSProtocolMajorVersion'] == 6:
                logging.info("Received V6 response")
		response = readKmsResponseV6(data)
	else:
                logging.info("Unhandled response version: %d.%d" % (config['KMSProtocolMajorVersion'], config['KMSProtocolMinorVersion']))
		logging.info("I'm not even sure how this happened...")
	return response

def readKmsResponseV4(data, request):
	response = kmsRequestV4.ResponseV4(data)
	hashed = kmsRequestV4(data, config).generateHash(bytearray(str(response['response'])))
	logging.info("Response Hash has expected value: ", hashed == response['hash'])
	return response

def readKmsResponseV5(data):
	response = kmsRequestV5.ResponseV5(data)
	decrypted = kmsRequestV5(data, config).decryptResponse(response)
	return decrypted

def readKmsResponseV6(data):
	response = kmsRequestV6.ResponseV5(data)
	decrypted = kmsRequestV6(data, config).decryptResponse(response)
	message = decrypted['message']
	return message

if __name__ == "__main__":
	main()
